rust/datafusion/src/datasource/parquet.rs

```diff
@@ -0,0 +1,586 @@
+            let converted_buffer: Vec<A::Native> =
+                read_buffer.into_iter().map(|v| v.into()).collect();
+            if values_read == levels_read {
+                builder.append_slice(&converted_buffer[0..values_read])?;
+            } else {
+                let mut value_index = 0;
+                for i in 0..def_levels.len() {
+                    if def_levels[i] != 0 {
+                        builder.append_value(converted_buffer[value_index].into())?;
+                        value_index += 1;
+                    } else {
+                        builder.append_null()?;
+                    }
+                }
+            }
+            Ok(Arc::new(builder.finish()))
+        } else {
+            let (values_read, _) =
+                self.read_batch(batch_size, None, None, &mut read_buffer)?;
+
+            let mut builder = PrimitiveBuilder::<A>::new(values_read);
+            let converted_buffer: Vec<A::Native> =
+                read_buffer.into_iter().map(|v| v.into()).collect();
+            builder.append_slice(&converted_buffer[0..values_read])?;
+            Ok(Arc::new(builder.finish()))
+        }
+    }
+}
+
+impl ParquetFile {
+    pub fn open(file: File, projection: Option<Vec<usize>>) -> Result<Self> {
+        let reader = SerializedFileReader::new(file)?;
+
+        let metadata = reader.metadata();
+        let schema =
+            parquet_to_arrow_schema(metadata.file_metadata().schema_descr_ptr())?;
+
+        // even if we aren't referencing structs or lists in our projection, column reader
+        // indexes will be off until we have support for nested schemas
+        for i in 0..schema.fields().len() {
+            match schema.field(i).data_type() {
+                DataType::List(_) => {
+                    return Err(ExecutionError::NotImplemented(
+                        "Parquet datasource does not support LIST".to_string(),
+                    ));
+                }
+                DataType::Struct(_) => {
+                    return Err(ExecutionError::NotImplemented(
+                        "Parquet datasource does not support STRUCT".to_string(),
+                    ));
+                }
+                _ => {}
+            }
+        }
+
+        let projection = match projection {
+            Some(p) => p,
+            None => {
+                let mut p = Vec::with_capacity(schema.fields().len());
+                for i in 0..schema.fields().len() {
+                    p.push(i);
+                }
+                p
+            }
+        };
+
+        let projected_schema = schema_projection(&schema, &projection)?;
+
+        Ok(ParquetFile {
+            reader: reader,
+            row_group_index: 0,
+            projection_schema: projected_schema,
+            projection,
+            batch_size: 64 * 1024,
+            current_row_group: None,
+            column_readers: vec![],
+        })
+    }
+
+    fn load_next_row_group(&mut self) -> Result<()> {
```

Comment:
can we have a test for this case as well? i.e., loading multiple row groups.

Reply:
done